Add support for timing C++ snippets.

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47865 Enable callgrind collection for C++ snippets
* **#47864 Add support for timing C++ snippets.**
* #47863 Rework compat bindings.
* #47862 Add option for cpp_extensions to compile standalone executable
* #47861 Expose CXX_FLAGS through __config__

Differential Revision: [D25199262](https://our.internmc.facebook.com/intern/diff/D25199262)

Commit:
Add support for timing C++ snippets.

[ghstack-poisoned]

Commit:
Update on "Add support for timing C++ snippets."




[ghstack-poisoned]

ezyang commented on torch/utils/benchmark/utils/timer.py:
```diff
@@ -158,15 +210,15 @@ def __init__(
+        stmt = (stmt[1:] if stmt and stmt[0] == "\n" else stmt).rstrip()
         setup = textwrap.dedent(setup)
-        setup = (setup[1:] if setup[0] == "\n" else setup).rstrip()
+        setup = (setup[1:] if setup and setup[0] == "\n" else setup).rstrip()
```
I agree the new code is better than the old code, but this is my skeptical face 🤔

ezyang commented on torch/utils/benchmark/utils/cpp_jit.py:
```diff
@@ -1,15 +1,24 @@
 SOURCE_ROOT = os.path.split(os.path.abspath(__file__))[0]
+BUILD_ROOT = os.path.join(
+    torch._appdirs.user_cache_dir(appname="benchmark_utils_jit"),
+    f"build_{uuid.uuid4()}".replace("-", "")
```
Calcing the build uuid only once at module import seems pretty skeevy to me...

robieta replied:
I was about to write a big paragraph on why it's not, but then I thought "that big paragraph should be a comment!".

ezyang replied:
Hmm... well this makes sense, but it raises a new question: do you *trust* `cpp_extension` s recompilation cache? When it comes to benchmarking, I'm not sure I would...

robieta replied:
I have two safeguards:
1) The `uuid` in the build root, so it won't mix `libtorch` across PyTorch versions.
2) I put the hash of the source in the name before I pass it to `cpp_extension`

So really `cpp_jit.py` guarantees cache correctness; it just has to make sure it doesn't also vary something that `cpp_extension` caches on. (i.e. problems will manifest as excessive recompilation, but never correctness issues.)

Want me to add a note to that effect?

ezyang commented on test/benchmark_utils/test_benchmark_utils.py:
```diff
@@ -162,6 +162,16 @@ def test_timer(self):
+            setup="torch::Tensor x = torch::empty({1});",
+            language=benchmark_utils.Language.CPP,
+        )
+        t = timer.timeit(10)
```
Is there really any good reason to run this for 10s? For the purpose of a test haha

robieta replied:
It's 10 iterations. `timer.blocked_autorange(min_run_time=10)` would be 10 seconds.

ezyang replied:
Converse question, then, is why is this a slow test :) (compile time?)

robieta replied:
Yup, compile time.

Commit:
Update on "Add support for timing C++ snippets."




[ghstack-poisoned]

robieta commented on torch/utils/benchmark/utils/common.py:
```diff
@@ -48,7 +48,7 @@ def title(self) -> str:
     def setup_str(self) -> str:
         return (
-            "" if self.setup == "pass"
+            "" if (self.setup == "pass" or not self.setup)
```
I don't think I can, since it's a user defined string. It's kind of like an `Optional[str]`, except that `timeit.Timer` doesn't allow an empty `setup`. I agree it's slightly curious.

ezyang replied:
oh, pass isn't a keyword, it's literally the Python code "pass". That's curious, but I guess matching timeit.Timer takes precedence here.

Commit:
Update on "Add support for timing C++ snippets."




[ghstack-poisoned]

Commit:
Update on "Add support for timing C++ snippets."




[ghstack-poisoned]

ezyang commented on torch/utils/benchmark/utils/timer.py:
```diff
@@ -135,21 +176,32 @@ def __init__(
+        # We copy `globals` to prevent mutations from leaking.
+        # (For instance, `eval` adds the `__builtins__` key)
+        self._globals = dict(globals or {})
+        if language in (Language.PYTHON, "py", "python"):
```
nit: would be cleaner to have a function that takes care of normalizing string to the Language Enum, and then use that. I don't think it's a big deal in this patch because the logic here is still centralized, but the moment you want to have a second interface that takes in string languages, you're gonna need this refactor.

ezyang commented on torch/utils/benchmark/utils/cpp_jit.py:
```diff
@@ -66,3 +89,55 @@ def get_compat_bindings() -> CallgrindModuleType:
+    with LOCK:
+        if not os.path.exists(BUILD_ROOT):
+            os.makedirs(BUILD_ROOT)
+            atexit.register(shutil.rmtree, BUILD_ROOT)
```
https://github.com/pytorch/pytorch/pull/47428/files#r521094904

robieta replied:
I'm implicitly relying on `torch._appdirs.user_cache_dir` having similar kinds of litter safety. (That's what you would get from vanilla `cpp_extension` anyway) On my machine it's `~/.cache/...`, for instance. And if you're still unconvinced it's no trouble to change it to use `tempfile` instead.

ezyang replied:
https://askubuntu.com/questions/102046/is-it-okay-to-delete-the-cache-folder seems to suggest that `~/.cache` doesn't get automatically cleaned.

ezyang reviewed (approved):
okey dokey

Commit:
Update on "Add support for timing C++ snippets."




[ghstack-poisoned]

Commit:
Update on "Add support for timing C++ snippets."




[ghstack-poisoned]

Commit:
Update on "Add support for timing C++ snippets."


Differential Revision: [D25199262](https://our.internmc.facebook.com/intern/diff/D25199262)

[ghstack-poisoned]

Commit:
Update on "Add support for timing C++ snippets."


Differential Revision: [D25199262](https://our.internmc.facebook.com/intern/diff/D25199262)

[ghstack-poisoned]

Commit:
Update on "Add support for timing C++ snippets."


Differential Revision: [D25199262](https://our.internmc.facebook.com/intern/diff/D25199262)

[ghstack-poisoned]

Commit:
Update on "Add support for timing C++ snippets."


Differential Revision: [D25199262](https://our.internmc.facebook.com/intern/diff/D25199262)

[ghstack-poisoned]
